File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        l_wall = r_wall = 0
        n = len(height)
        
        max_left = [0] * n
        max_right = [0] * n

        for i in range(n):
            max_left[i] = l_wall
            l_wall = max(l_wall,height[i])
        
        for j in range(n-1,-1,-1):
            max_right[j] = r_wall
            r_wall = max(r_wall,height[j])

        summ = 0
        for i in range(n):
            pot = min(max_left[i],max_right[i])
            summ += max(0,pot - height[i])
        return summ

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;

        int leftWall = 0;
        int rightWall = 0;

        int[] leftMax = new int[n];
        int[] rightMax = new int[n];

        for (int i = 0; i < n; i++) {
            leftMax[i] = leftWall;
            leftWall = Math.max(leftWall, height[i]);
        }

        for (int i = n - 1; i >= 0; i--) {
            rightMax[i] = rightWall;
            rightWall = Math.max(rightWall, height[i]);
        }

        int summ = 0;

        for (int i = 0; i < n; i++) {
            int pot = Math.min(leftMax[i], rightMax[i]);
            summ += Math.max(0, pot - height[i]);
        }

        return summ;
    }
}